specifications/filter.md

```diff
@@ -0,0 +1,307 @@
+# The LoopBack 4 Filter Specification - Draft 1
+
+## Abstract
+
+This document describes the LoopBack 4 Filter Specification; A JSON-based
+[[RFC7159](https://tools.ietf.org/html/rfc7159)] data query language that
+enables rich, interopable and protocol-agnostic data query filters across
+multiple datastores.
+
+## Status of This Document
+
+This is a draft document.
+
+## Copyright Notice
+
+Copyright (c) 2021 The LoopBack 4 Filter Specification Authors.
+
+## Table of Contents
+
+- 1 [Introduction](#1-introduction)
+
+  - 1.1 [Notational Conventions](#11-notational-conventions)
+  - 1.2 [Terminology](#12-terminology)
+  - 1.3 [Motivation, Prior Work and Scope](#13-motivation-prior-work-and-scope)
+
+- 2 [Property Identifier](#2-property-identifier)
+
+- 3 [Filter Syntax](#3-filter-syntax)
+
+  - 3.1 [Fields Sub-Filter](#31-fields-sub-filter)
+  - 3.2 [Order Sub-Filter](#32-order-sub-filter)
+
+## 1 Introduction
+
+### 1.1 Notational Conventions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [[RFC2119](https://tools.ietf.org/html/rfc2119)].
+
+The grammatical rules in this document are to be interpreted as described in
+[[RFC5234](https://tools.ietf.org/html/rfc5234)].
+
+This document makes reference to grammatical rules defined in
+[[RFC7159](https://tools.ietf.org/html/rfc7159)].
+
+### 1.2 Terminology
+
+This specification defines the following terms:
+
+- _Compliance Matrix_: A standardised, grouped list of Filter Features used to
+  provide a higher-level overlay to gauge general interopability of features
+  without requiring deep knowledge on the subject matter.
+- _Connector_: A code implementation that interfaces with a Datastore that MUST
+  accept a Normalized Filter for data querying against the Datastore.
+- _Datastore_: A code implementation that manages the state of Connectors.
```

Comment:
We should generalise these to something along the lines of "Data Producer" as this specification should be scoped to only passing of filters from Data Consumer to Data Producer. What exactly is the Data Producer and Data Consumer should be up to the implementations to decide.

For LoopBack Juggler and LoopBack 4, we should informally document that the Connector accepts a normalized filter in line with a specific compliance matrix, the LB4 Repository (and inherently the REST API) accepts un-normalized filter with specific compliance matrix, etc.